proxy.py

```diff
@@ -2,225 +2,145 @@
-        '3': 'ycm_extra_conf_sdk3.py',
-    }
-    sdk_mapping = {
-        '2': settings.PEBBLE_SDK2,
-        '3': settings.PEBBLE_SDK3,
+    result = ycm_helpers.spinup(content)
+    result['ws_port'] = settings.PORT
+    return jsonify(result)
+
+
+def server_ws(process_uuid):
+    ws_commands = {
+        'completions': ycm_helpers.get_completions,
+        'errors': ycm_helpers.get_errors,
+        'goto': ycm_helpers.go_to,
+        'create': ycm_helpers.create_file,
+        'delete': ycm_helpers.delete_file,
+        'ping': ycm_helpers.ping
     }
 
-    with open(settings_path, "w") as f:
-        with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'ycm_conf', conf_mapping[sdk_version])) as template:
-            f.write(template.read().format(sdk=sdk_mapping[sdk_version], here=root_dir, stdlib=settings.STDLIB_INCLUDE_PATH))
-
-    try:
-        if 'aplite' in platforms:
-            ycm = YCM(filesync, 'aplite')
-            ycm.wait()
-            ycm.apply_settings(settings_path)
-            ycms.ycms['aplite'] = ycm
-
-        if 'basalt' in platforms:
-            ycm = YCM(filesync, 'basalt')
-            ycm.wait()
-            ycm.apply_settings(settings_path)
-            ycms.ycms['basalt'] = ycm
-    except Exception as e:
-        print "Failed to spawn ycm with root_dir %s" % root_dir
-        print traceback.format_exc()
-        return jsonify(success=False, error=str(e)), 500
-
-    # Keep track of it
-    this_uuid = str(uuid.uuid4())
-    mapping[this_uuid] = ycms
-    print mapping
-    print "spinup complete (%s); %s -> %s" % (platforms, this_uuid, root_dir)
-    # victory!
-    return jsonify(success=True, uuid=this_uuid)
-
-
-@app.route('/ycm/<process_uuid>/completions', methods=['POST'])
-def get_completions(process_uuid):
-    if process_uuid not in mapping:
-        return "Not found", 404
-    ycms = mapping[process_uuid]
-    data = request.get_json(force=True)
-    if 'patches' in data:
-        ycms.filesync.apply_patches(data['patches'])
-    completions = collections.OrderedDict()
-    completion_start_column = None
-    for platform, ycm in sorted(ycms.ycms.iteritems(), reverse=True):
-        ycm.parse(data['file'], data['line'], data['ch'])  # TODO: Should we do this here?
-        platform_completions = ycm.get_completions(data['file'], data['line'], data['ch'])
-        if platform_completions is not None:
-            completion_start_column = platform_completions['completion_start_column']
-            for completion in platform_completions['completions']:
-                if completion['insertion_text'] in completions:
+    # Get the WebSocket from the request context
+    server_ws = request.environ.get('wsgi.websocket', None)
+    if server_ws is None:
+        return "websocket endpoint", 400
+
+    alive = [True]
+    def do_recv(receive, send, error):
+        try:
+            while alive[0]:
+                raw = receive()
+                id = -1
```

Comment:
Well, it doesn't get overwritten if packet.loads() throws an exception, or if packet['_ws_message_id'] is undefined, in which case error(id, 'invalid pakcet') is called with id=-1 and the message sent back to cloudpebble will have an id of -1. 

-1 doesn't have any special significance except for being less than 0, so cloudpebble would never expect an id of -1. I suppose I could set it to None though, or just not send an error message back at all, or send one back but with no id field.